### PR TITLE
Add integration test cases for checking the behavior of generated code

### PR DESCRIPTION
**Pull request by shiraji**
This test cases are useful for when we are integrating Kotlin base
dispatcher. Does test cases make sure the code we generated are
correctly working. The test cases doesn't care what kind of code we
generate.

### Concern

* ~~I created test-v13 module but it seems this test case doesn't care v13 or not. We may need to add them to test module.~~
* This is the first time I used Roboletric+PowerMockito. I may not understand them incorrectly. Please check the test cases carefully.
* Test cases except all annotation classes may not be useful.

Thanks

### CODE REVIEW

**Comment by hotchemi** [comment]
Thx! let me take a look.

File: test-v13/src/test/java/permissions/dispatcher/test_v13/FragmentWithAllAnnotationsKtPermissionsDispatcherTest.kt [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+import permissions.dispatcher.PermissionUtils
+import org.powermock.modules.junit4.rule.PowerMockRule
+import org.junit.Rule
+import permissions.dispatcher.test.*
+
+
+@Suppress("IllegalIdentifier")
+@RunWith(RobolectricTestRunner::class)
+@PowerMockIgnore("org.mockito.*", "org.robolectric.*", "android.*")
+@PrepareForTest(PermissionUtils::class)
```
Comment (shiraji) [inline test-v13/src/test/java/permissions/dispatcher/test_v13/FragmentWithAllAnnotationsKtPermissionsDispatcherTest.kt]: All test cases mock `PermissionUtils` class but I start feeling this should not be mock...let me change where to mock. I think `PermissionChecker` should be the class to mock

**Comment by shiraji** [comment]
#340 is blocking this PR.

File: test-v13/build.gradle [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    packagingOptions {
+        exclude 'mockito-extensions/org.mockito.plugins.MockMaker'
+    }
+}
+
+dependencies {
+    compile fileTree(dir: 'libs', include: ['*.jar'])
+
+    compile (project(':library'))
+    provided project(':processor')
```
Comment (hotchemi) [inline test-v13/build.gradle]: kpat and provided are both needed right?

File: test-v13/build.gradle [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    kapt project(':processor')
+
+    compile "org.jetbrains.kotlin:kotlin-stdlib:$KOTLIN_VERSION"
+    compile 'com.android.support:appcompat-v7:25.3.1'
```
Comment (hotchemi) [inline test-v13/build.gradle]: can we use `SUPPORT_LIBRARY_VERSION`?

File: test-v13/proguard-rules.pro [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# Add project specific ProGuard rules here.
```
Comment (hotchemi) [inline test-v13/proguard-rules.pro]: is this file needed?

File: test/proguard-rules.pro [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# Add project specific ProGuard rules here.
```
Comment (hotchemi) [inline test/proguard-rules.pro]: is this file needed?

File: test/build.gradle [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    compile fileTree(dir: 'libs', include: ['*.jar'])
+
+    compile (project(':library')) {
+        exclude module: "support-v13"
+    }
+    provided project(':processor')
+    kapt project(':processor')
+
+    compile "org.jetbrains.kotlin:kotlin-stdlib:$KOTLIN_VERSION"
+    compile 'com.android.support:appcompat-v7:25.3.1'
```
Comment (hotchemi) [inline test/build.gradle]: can we use SUPPORT_LIBRARY_VERSION?

**Comment by hotchemi** [comment]
one more plus, I checked carefully but I can't assure we've covered all test cases in this PR. So it'd be fine to add coverage report service integration to detect uncovered part of code base(of course that's not necessary for this PR!).

**Comment by shiraji** [comment]
> we're gonna add Kotlin support how we can test it?

These are the class written in Kotlin

* ActivityWithAllAnnotationsKt
* FragmentWithAllAnnotationsKt
* SupportFragmentWithAllAnnotationsKt

Are you going to change the interface of calling the method? Then, we can change the test cases.

> coverage

One big problem is this test cases are testing generated code, not existed code. I have no idea there is a library for that.

File: test-v13/build.gradle [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        targetCompatibility JavaVersion.VERSION_1_7
+    }
+    lintOptions {
+        abortOnError false
+    }
+    sourceSets {
+        main.java.srcDirs += "src/main/kotlin"
+    }
+    packagingOptions {
+        exclude 'mockito-extensions/org.mockito.plugins.MockMaker'
```
Comment (shiraji) [inline test-v13/build.gradle]: When I searched to using both Mockito and PowerMockito, I need to add this line. Let me check!

File: test-v13/build.gradle (hunk 2 above)
Comment (shiraji) [inline test-v13/build.gradle]: Yea, as we find that in #328

File: test-v13/proguard-rules.pro (hunk 4 above)
Comment (shiraji) [inline test-v13/proguard-rules.pro]: It's default one. We can delete proguard settings.

File: test-v13/build.gradle (hunk 7 above)
Comment (aurae) [inline test-v13/build.gradle]: Both libraries are packaging a file of that name inside their JAR/AAR, so there might be a conflict when packaging the merged APK.

Comment (shiraji) [inline test-v13/build.gradle]: Yea, but without this setting, I can run `./gradlew check`, so let's delete this.

File: test-v13/src/test/java/permissions/dispatcher/test_v13/Extention.kt [hunk 8]
```diff
@@ -0,0 +1,33 @@
+package permissions.dispatcher.test
```
Comment (aurae) [inline test-v13/src/test/java/permissions/dispatcher/test_v13/Extention.kt]: Please rename to `Extensions.kt`

File: test/src/test/java/permissions/dispatcher/test/Extention.kt [hunk 9]
```diff
@@ -0,0 +1,42 @@
+package permissions.dispatcher.test
```
Comment (aurae) [inline test/src/test/java/permissions/dispatcher/test/Extention.kt]: Please rename to `Extensions.kt`

**Comment by shiraji** [comment]
Thanks @aurae I renamed them.

**Comment by hotchemi** [comment]
https://github.com/hotchemi/PermissionsDispatcher/pull/339#issuecomment-314376678

ic. so let's keep existing test cases and have a Jacoco integration to monitor a coverage!

**Comment by hotchemi** [comment]
Is it ready to merge?

**Comment by shiraji** [comment]
Not yet. i will rename file names and squash commits

**Comment by shiraji** [comment]
Renamed Exten**t**ions.kt to Exten**s**ions.kt and squashed commits.

Now it ready to merge this PR!

**Comment by hotchemi** [comment]
@shiraji thx! About Jacoco I'll work on it!
